src/curve/encoding.rs

```diff
@@ -0,0 +1,488 @@
+            CtOption::new(
+                AffinePoint::identity(),
+                infinity_flag_set & // Infinity flag should be set
+                    (!sort_flag_set) & // Sort flag should not be set
+                    x.is_zero(), // The x-coordinate should be zero
+            )
+            .or_else(|| {
+                // Recover a y-coordinate given x by y = sqrt(x^3 + x + B)
+                ((x.square() * x) + x + B).sqrt().and_then(|y| {
+                    // Switch to the correct y-coordinate if necessary.
+                    let y = Fp6::conditional_select(
+                        &y,
+                        &-y,
+                        y.lexicographically_largest() ^ sort_flag_set,
+                    );
+
+                    CtOption::new(
+                        AffinePoint {
+                            x,
+                            y,
+                            infinity: infinity_flag_set,
+                        },
+                        !infinity_flag_set, // Infinity flag should not be set
+                    )
+                })
+            })
+        })
+    }
+
+    /// Copies the bytes of this `CompressedPoint`.
+    pub fn to_bytes(&self) -> [u8; 49] {
+        self.0
+    }
+
+    /// Views this `CompressedPoint` as an array of bytes.
+    pub fn as_bytes(&self) -> &[u8; 49] {
+        &self.0
+    }
+
+    /// Interprets the provided bytes as a `CompressedPoint`.
+    /// This does not check the validity of the input, and may result
+    /// in failure when calling `to_affine()`.
+    pub fn from_bytes(bytes: &[u8; 49]) -> Self {
+        Self(*bytes)
+    }
+}
+
+/// A uncompressed point, storing the `x` and `y` coordinates
+/// of a point, along an extra byte storing metadata to be
+/// used for decompression.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct UncompressedPoint(pub [u8; 97]);
+
+impl Default for UncompressedPoint {
+    fn default() -> Self {
+        AffinePoint::identity().to_uncompressed()
+    }
+}
+
+impl ConstantTimeEq for UncompressedPoint {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.as_bytes().ct_eq(other.as_bytes())
+    }
+}
+
+impl UncompressedPoint {
+    /// Converts an `AffinePoint` to an `UncompressedPoint`
+    pub(crate) fn from_affine(point: &AffinePoint) -> Self {
+        let mut bytes = [0; 97];
+
+        bytes[0..48].copy_from_slice(
+            &Fp6::conditional_select(&point.x, &Fp6::zero(), point.infinity).to_bytes()[..],
+        );
+        bytes[48..96].copy_from_slice(
+            &Fp6::conditional_select(&point.y, &Fp6::zero(), point.infinity).to_bytes()[..],
+        );
+
+        // Is this point at infinity? If so, set the most significant bit of the last byte.
+        bytes[96] |= u8::conditional_select(&0u8, &(1u8 << 7), point.infinity);
+
+        Self(bytes)
+    }
+
+    /// Attempts to convert an `UncompressedPoint` to an `AffinePoint`
+    /// The resulting point is ensured to be on the curve, but is
```

Comment:
Is this point actually ensured to be on the curve? It seems we just recover x and y as given.

Reply:
This is an incorrect leftover from some previous PR. Indeed, for `UncompressedPoint`, as there is no `y` recovery, the additional curve check must be done outside of this method. This is done on the `AffinePoint` method `from_uncompressed`, on which other points coordinate methods are based. Conversion from a `CompressedPoint` already includes this check and should hence only check that we are on the correct subgroup. I'll correct it, thanks!